python/pyspark/sql/types.py

```diff
@@ -876,30 +894,86 @@ def __init__(
         self.dataType = dataType
         self.nullable = nullable
         self.metadata = metadata or {}
+        self._collationMetadata: Optional[Dict[str, str]] = None
 
     def simpleString(self) -> str:
         return "%s:%s" % (self.name, self.dataType.simpleString())
 
     def __repr__(self) -> str:
         return "StructField('%s', %s, %s)" % (self.name, self.dataType, str(self.nullable))
 
+    def __eq__(self, other: Any) -> bool:
+        # since collationMetadata is lazy evaluated we should not use it in equality check
```

Comment:
I would argue that just using `self.__dict__ == other.__dict__` is also dangerous and yet this is the current implementation of equals in all data types. This was only done in order for assert to pass in the tests

Reply:
For me this also points to the question if we should actually have a _collationMetadata member. It is not part of the state of the struct field, and the problem that can come from caching this value is that we might forget to update it. So do we better drop the caching? And we should have a test for
Create schema -> serialize json -> update collation id -> serialize json again

Reply:
I agree, doing this for every json call is annoying but since we don't have the security of immutable data like in scala perhaps we should avoid lazy eval